File: selenium-test-suite/tests/sdk-audit-hidden-content-thresholds.test.js
/**
 * Fake Hidden Content Audit Tests (Selenium)
 *
 * Mirrors the Playwright fake-hidden-content-audit.spec.js using Selenium WebDriver.
 */

const { By } = require('selenium-webdriver');
const { AccessFlowSDK, SeleniumDriver } = require('@acsbe/accessflow-sdk');
const { createDriver, navigateTo } = require('../helpers/driver');
const fs = require('fs');
const path = require('path');

AccessFlowSDK.init({
  apiKey:
    process.env.AF_NODE_PACKAGE_KEY ||
    process.env.AF_Node_Package_Key ||
    'flow-1TmAHDBGjDGX4aadlA0000VdvvxVnhZfP2',
});

const configPath = path.join(__dirname, '..', 'accessflow.config.json');
let config = {
  issuesFoundThreshold: { extreme: 0, high: 0, medium: 0, low: 0 },
  localCheck: true,
};
if (fs.existsSync(configPath)) {
  config = JSON.parse(fs.readFileSync(configPath, 'utf8'));
}

let driver;

beforeAll(async () => {
  driver = await createDriver();
});

afterAll(async () => {
  if (driver) await driver.quit();
});

describe('Fake Hidden Content Audit Tests', () => {
  beforeEach(async () => {
    await navigateTo(driver, '/');
    await driver.sleep(1000);
  });

  test('should perform audits across multiple states on fake hidden content page and verify thresholds', async () => {
    const auditResults = [];
    const sdk = new AccessFlowSDK(new SeleniumDriver(driver));

    const performAudit = async (context) => {
      console.log(`Starting audit for context: ${context}`);
      const report = await sdk.audit();
      auditResults.push({ context, report });

      if (config.localCheck && report) {
        const issuesFound = report.numberOfIssuesFound || {};
        const thresholds = config.issuesFoundThreshold;
        let thresholdExceeded = false;

        for (const severity of ['extreme', 'high', 'medium', 'low']) {
          const count = issuesFound[severity] || 0;
          const limit =
            thresholds[severity] !== undefined ? thresholds[severity] : Infinity;

          if (count > limit) {
            console.log(
              `[Threshold Violation] ${severity}: ${count} > ${limit} (limit)`,
            );
            thresholdExceeded = true;
          }
        }

        if (context === 'Initial Failure Page') {
          console.log(
            `[${context}] Threshold exceeded: ${thresholdExceeded} (this page displays failure code examples, not actual violations)`,
          );
        }
      }

      console.log(`Audit completed for ${context}`);
    };

    // Navigate to failure page
    await navigateTo(driver, '/errors/fake-hidden-content_failure');
    await driver.sleep(1000);
    const currentUrl = await driver.getCurrentUrl();
    expect(currentUrl).toContain('errors/fake-hidden-content_failure');

    // State 1: Initial Load (Failure Page)
    await performAudit('Initial Failure Page');

    // State 2: Open Modal (Notifications)
    try {
      const notificationButton = await driver.findElement(
        By.css('button:has(svg[data-testid="NotificationsIcon"])'),
      );
      if (await notificationButton.isDisplayed()) {
        await notificationButton.click();
        await driver.sleep(500);
        await performAudit('Notifications Modal Open');

        const body = await driver.findElement(By.css('body'));
        await body.sendKeys('\uE00C');
        await driver.sleep(500);
      }
    } catch {
      // Notification button not found
    }

    // State 3: Expand/Collapse Sidebar
    try {
      const closeSidebarButton = await driver.findElement(
        By.css('button:has(svg[data-testid="ChevronLeftIcon"])'),
      );
      if (await closeSidebarButton.isDisplayed()) {
        await closeSidebarButton.click();
        await driver.sleep(500);
        await performAudit('Sidebar Collapsed');

        try {
          const openSidebarButton = await driver.findElement(
            By.css('button:has(svg[data-testid="MenuIcon"])'),
          );
          if (await openSidebarButton.isDisplayed()) {
            await openSidebarButton.click();
          }
        } catch {
          // Menu button not found
        }
      }
    } catch {
      // Sidebar toggle not found
    }

    // State 4: Navigate to Success Page
    await navigateTo(driver, '/errors/fake-hidden-content_success');
    await driver.sleep(1000);
    const successUrl = await driver.getCurrentUrl();
    expect(successUrl).toContain('errors/fake-hidden-content_success');

    await performAudit('Success Page');

    // Verification
    console.log('Total audits performed:', auditResults.length);
    expect(auditResults.length).toBeGreaterThan(1);

    const violationCounts = auditResults.map((r) => {
      let violationsCount = 0;
      if (r.report && r.report.ruleViolations) {
        violationsCount = Object.values(r.report.ruleViolations).reduce(
          (acc, rule) => acc + (rule.numberOfOccurrences || 0),
          0,
        );
      } else if (r.report && r.report.violations) {
        violationsCount = r.report.violations.length;
      }
      return { context: r.context, violations: violationsCount };
    });

    console.log(
      'Violation counts by context:',
      JSON.stringify(violationCounts, null, 2),
    );
  });
});

File: java-selenium-tests/src/test/java/com/accessflow/SdkAuditHiddenContentThresholdsTest.java
package com.accessflow;

import com.acsbe.accessflow.AccessFlowSDK;
import com.acsbe.accessflow.AccessFlowTeardown;
import com.acsbe.accessflow.SeleniumDriver;
import org.junit.jupiter.api.*;
import org.openqa.selenium.*;

import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Fake Hidden Content Audit / Threshold Tests (Selenium).
 * Mirrors Node sdk-audit-hidden-content-thresholds.test.js.
 */
public class SdkAuditHiddenContentThresholdsTest {

    private WebDriver driver;
    private static Map<String, Object> config;

    @BeforeAll
    @SuppressWarnings("unchecked")
    static void loadConfig() {
        try (InputStream is = SdkAuditHiddenContentThresholdsTest.class.getResourceAsStream("/accessflow.config.json")) {
            if (is != null) {
                String json = new String(is.readAllBytes(), StandardCharsets.UTF_8);
                config = parseJsonToMap(json);
            }
        } catch (Exception e) {
            System.err.println("Failed to load accessflow.config.json: " + e.getMessage());
        }
        if (config == null) {
            config = Map.of(
                    "issuesFoundThreshold", Map.of("extreme", 0, "high", 0, "medium", 0, "low", 0),
                    "localCheck", true
            );
        }
    }

    @BeforeEach
    void setup() {
        driver = DriverHelper.createDriver();
    }

    @AfterEach
    void teardown() {
        if (driver != null) driver.quit();
    }

    @AfterAll
    static void finalizeReports() {
        LocalReportCollector.write();
        AccessFlowTeardown.finalizeReports();
    }

    private AccessFlowSDK createSdk() {
        String apiKey = DriverHelper.getApiKey();
        return apiKey != null && !apiKey.isEmpty()
                ? new AccessFlowSDK(new SeleniumDriver(driver), apiKey)
                : new AccessFlowSDK(new SeleniumDriver(driver));
    }

    @Test
    @DisplayName("Should perform audits across states and verify thresholds")
    void testAuditAcrossStatesAndVerifyThresholds() {
        List<Map<String, Object>> auditResults = new ArrayList<>();

        DriverHelper.navigateTo(driver, "/errors/fake-hidden-content_failure");
        DriverHelper.sleep(1000);
        assertTrue(driver.getCurrentUrl().contains("errors/fake-hidden-content_failure"));

        AccessFlowSDK sdk = createSdk();

        Map<String, Object> initialReport = sdk.audit();
        auditResults.add(Map.of("context", "Initial Failure Page", "report", initialReport != null ? initialReport : Map.of()));
        AccessFlowTeardown.recordAudit(driver.getCurrentUrl(), initialReport);
        LocalReportCollector.add(driver.getCurrentUrl(), initialReport);

        try {
            WebElement notificationBtn = driver.findElement(
                    By.xpath("//button[.//svg[@data-testid='NotificationsIcon']]"));
            if (notificationBtn.isDisplayed()) {
                notificationBtn.click();
                DriverHelper.sleep(500);
                Map<String, Object> modalReport = sdk.audit();
                auditResults.add(Map.of("context", "Notifications Modal Open", "report", modalReport != null ? modalReport : Map.of()));
                AccessFlowTeardown.recordAudit(driver.getCurrentUrl(), modalReport);
                LocalReportCollector.add(driver.getCurrentUrl(), modalReport);
                driver.findElement(By.cssSelector("body")).sendKeys(Keys.ESCAPE);
                DriverHelper.sleep(500);
            }
        } catch (NoSuchElementException ignored) {}

        try {
            WebElement closeSidebar = driver.findElement(
                    By.xpath("//button[.//svg[@data-testid='ChevronLeftIcon']]"));
            if (closeSidebar.isDisplayed()) {
                closeSidebar.click();
                DriverHelper.sleep(500);
                Map<String, Object> collapsedReport = sdk.audit();
                auditResults.add(Map.of("context", "Sidebar Collapsed", "report", collapsedReport != null ? collapsedReport : Map.of()));
                AccessFlowTeardown.recordAudit(driver.getCurrentUrl(), collapsedReport);
                LocalReportCollector.add(driver.getCurrentUrl(), collapsedReport);
                try {
                    WebElement openSidebar = driver.findElement(
                            By.xpath("//button[.//svg[@data-testid='MenuIcon']]"));
                    if (openSidebar.isDisplayed()) openSidebar.click();
                } catch (NoSuchElementException ignored) {}
            }
        } catch (NoSuchElementException ignored) {}

        DriverHelper.navigateTo(driver, "/errors/fake-hidden-content_success");
        DriverHelper.sleep(1000);
        assertTrue(driver.getCurrentUrl().contains("errors/fake-hidden-content_success"));

        Map<String, Object> successReport = sdk.audit();
        auditResults.add(Map.of("context", "Success Page", "report", successReport != null ? successReport : Map.of()));
        AccessFlowTeardown.recordAudit(driver.getCurrentUrl(), successReport);
        LocalReportCollector.add(driver.getCurrentUrl(), successReport);

        assertTrue(auditResults.size() > 1);
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> parseJsonToMap(String json) {
        // Minimal JSON parser — sufficient for flat/nested maps of primitives
        json = json.trim();
        if (!json.startsWith("{")) return Map.of();
        java.util.LinkedHashMap<String, Object> map = new java.util.LinkedHashMap<>();
        json = json.substring(1, json.lastIndexOf("}")).trim();
        for (String entry : splitJsonEntries(json)) {
            String[] kv = entry.split(":", 2);
            String key = kv[0].trim().replaceAll("^\"|\"$", "");
            String val = kv[1].trim();
            if (val.startsWith("{")) {
                map.put(key, parseJsonToMap(val));
            } else if (val.equals("true")) {
                map.put(key, true);
            } else if (val.equals("false")) {
                map.put(key, false);
            } else if (val.startsWith("\"")) {
                map.put(key, val.replaceAll("^\"|\"$", ""));
            } else {
                try { map.put(key, Integer.parseInt(val)); }
                catch (NumberFormatException e) {
                    try { map.put(key, Double.parseDouble(val)); }
                    catch (NumberFormatException e2) { map.put(key, val); }
                }
            }
        }
        return map;
    }

    private static List<String> splitJsonEntries(String json) {
        List<String> entries = new ArrayList<>();
        int depth = 0;
        int start = 0;
        for (int i = 0; i < json.length(); i++) {
            char c = json.charAt(i);
            if (c == '{') depth++;
            else if (c == '}') depth--;
            else if (c == ',' && depth == 0) {
                entries.add(json.substring(start, i).trim());
                start = i + 1;
            }
        }
        String last = json.substring(start).trim();
        if (!last.isEmpty()) entries.add(last);
        return entries;
    }
}
